pkg/controller/performanceprofile/performanceprofile_controller.go

```diff
@@ -103,54 +210,350 @@ func (r *ReconcilePerformanceProfile) Reconcile(request reconcile.Request) (reco
+		if hasFinalizer(instance, finalizer) {
+			removeFinalizer(instance, finalizer)
+			if err := r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{}, nil
+		}
 	}
 
-	// Check if this Pod already exists
-	found := &corev1.Pod{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
-		if err != nil {
+	// TODO: we need to check if all under performance profiles values != nil
+	// first we need to decide if each of values required and we should move the check into validation webhook
+	// for now let's assume that all parameters needed for assets scrips are required
+	if err := r.validatePerformanceProfileParameters(instance); err != nil {
+		// we do not want to reconcile again in case of error, because a user will need to update the PerformanceProfile anyway
+		klog.Errorf("failed to reconcile: %v", err)
+		return reconcile.Result{}, nil
+	}
+
+	// add finalizer
+	if !hasFinalizer(instance, finalizer) {
+		instance.Finalizers = append(instance.Finalizers, finalizer)
+		if err := r.client.Update(context.TODO(), instance); err != nil {
 			return reconcile.Result{}, err
 		}
 
-		// Pod created successfully - don't requeue
+		// we exit reconcile loop because we will have additional update reconcile
 		return reconcile.Result{}, nil
-	} else if err != nil {
+	}
+
+	// apply components
+	result, err := r.applyComponents(instance)
+	if err != nil {
+		klog.Errorf("failed to deploy components: %v", err)
 		return reconcile.Result{}, err
 	}
 
-	// Pod already exists - don't requeue
-	reqLogger.Info("Skip reconcile: Pod already exists", "Pod.Namespace", found.Namespace, "Pod.Name", found.Name)
+	// TODO: we need to update the status
+
+	if result != nil {
+		return *result, nil
+	}
+
 	return reconcile.Result{}, nil
 }
 
-// newPodForCR returns a busybox pod with the same name/namespace as the cr
-func newPodForCR(cr *performancev1alpha1.PerformanceProfile) *corev1.Pod {
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcilePerformanceProfile) applyComponents(profile *performancev1alpha1.PerformanceProfile) (*reconcile.Result, error) {
+	// get mutated machine config pool
+	mcp := machineconfigpool.New(profile)
+	if err := controllerutil.SetControllerReference(profile, mcp, r.scheme); err != nil {
+		return nil, err
 	}
-	return &corev1.Pod{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      cr.Name + "-pod",
-			Namespace: cr.Namespace,
-			Labels:    labels,
-		},
-		Spec: corev1.PodSpec{
-			Containers: []corev1.Container{
-				{
-					Name:    "busybox",
-					Image:   "busybox",
-					Command: []string{"sleep", "3600"},
-				},
-			},
-		},
+	mcpMutated, created, err := r.getMutatedMachineConfigPool(mcp)
```

Comment:
the ```created``` value here is confusing.  created=false when mcp exists. and created=true when mcp needs to be created.

can we call ```created``` something like ```nonCreated```, or reverse the meaning of ```created``` to reflect the status of the object being observed?

Reply:
After merging `pause` and `createOrUpdate` of MCP I dropped it, I just overthought it X_X